File: lib/agent.ts
// Tabular Q-learning agent (TS port of tradebot/agent.py).

import { N_ACTIONS, type State } from "./env";
import { Rng } from "./rng";

export interface AgentOptions {
  gamma?: number;
  alpha?: number;
  epsilon?: number;
  epsilonEnd?: number;
  epsilonDecay?: number;
  seed?: number;
}

export interface SerializedAgent {
  gamma: number;
  alpha: number;
  epsilon: number;
  epsilonEnd: number;
  epsilonDecay: number;
  seed: number;
  q: Record<string, number>;
}

function keyOf(s: State, a: number): string {
  return `${s[0]},${s[1]},${s[2]}|${a}`;
}

export class QAgent {
  gamma: number;
  alpha: number;
  epsilon: number;
  epsilonEnd: number;
  epsilonDecay: number;
  seed: number;
  q: Map<string, number>;
  private rng: Rng;

  constructor(opts: AgentOptions = {}) {
    this.gamma = opts.gamma ?? 0.97;
    this.alpha = opts.alpha ?? 0.1;
    this.epsilon = opts.epsilon ?? 1.0;
    this.epsilonEnd = opts.epsilonEnd ?? 0.05;
    this.epsilonDecay = opts.epsilonDecay ?? 0.995;
    this.seed = opts.seed ?? 42;
    this.q = new Map();
    this.rng = new Rng(this.seed);
  }

  private get(s: State, a: number): number {
    return this.q.get(keyOf(s, a)) ?? 0;
  }

  bestAction(s: State): number {
    let bestA = 0;
    let bestV = -Infinity;
    for (let a = 0; a < N_ACTIONS; a++) {
      const v = this.get(s, a);
      if (v > bestV) {
        bestV = v;
        bestA = a;
      }
    }
    return bestA;
  }

  act(s: State, explore: boolean): number {
    if (explore && this.rng.next() < this.epsilon) {
      return this.rng.nextInt(N_ACTIONS);
    }
    return this.bestAction(s);
  }

  update(s: State, a: number, r: number, sNext: State, done: boolean): void {
    const current = this.get(s, a);
    let target = r;
    if (!done) {
      let bestNext = -Infinity;
      for (let a2 = 0; a2 < N_ACTIONS; a2++) {
        const v = this.get(sNext, a2);
        if (v > bestNext) bestNext = v;
      }
      target = r + this.gamma * bestNext;
    }
    this.q.set(keyOf(s, a), current + this.alpha * (target - current));
  }

  decayEpsilon(): void {
    this.epsilon = Math.max(this.epsilonEnd, this.epsilon * this.epsilonDecay);
  }

  toJSON(): SerializedAgent {
    const q: Record<string, number> = {};
    this.q.forEach((v, k) => {
      q[k] = v;
    });
    return {
      gamma: this.gamma,
      alpha: this.alpha,
      epsilon: this.epsilon,
      epsilonEnd: this.epsilonEnd,
      epsilonDecay: this.epsilonDecay,
      seed: this.seed,
      q,
    };
  }

  static fromJSON(data: SerializedAgent): QAgent {
    const a = new QAgent({
      gamma: data.gamma,
      alpha: data.alpha,
      epsilon: data.epsilon,
      epsilonEnd: data.epsilonEnd,
      epsilonDecay: data.epsilonDecay,
      seed: data.seed,
    });
    for (const [k, v] of Object.entries(data.q)) a.q.set(k, v);
    return a;
  }
}

File: tradebot/agent.py
"""Tabular Q-learning agent.

State and action spaces are discrete (see ``env.py``), so we can keep the
Q-table as a plain dict. Serialization is pickle-based and fits in a few KB.
"""
from __future__ import annotations

import pickle
import random
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, Tuple

from .env import N_ACTIONS, State


@dataclass
class QAgent:
    gamma: float = 0.97
    alpha: float = 0.1
    epsilon: float = 1.0
    epsilon_end: float = 0.05
    epsilon_decay: float = 0.995
    seed: int = 42
    q: Dict[Tuple[State, int], float] = field(default_factory=dict)

    def __post_init__(self) -> None:
        self._rng = random.Random(self.seed)

    # ------------------------------------------------------------------
    # Policy
    # ------------------------------------------------------------------

    def _q(self, s: State, a: int) -> float:
        return self.q.get((s, a), 0.0)

    def best_action(self, s: State) -> int:
        best_a, best_v = 0, float("-inf")
        for a in range(N_ACTIONS):
            v = self._q(s, a)
            if v > best_v:
                best_v = v
                best_a = a
        return best_a

    def act(self, s: State, *, explore: bool = True) -> int:
        if explore and self._rng.random() < self.epsilon:
            return self._rng.randrange(N_ACTIONS)
        return self.best_action(s)

    # ------------------------------------------------------------------
    # Learning
    # ------------------------------------------------------------------

    def update(self, s: State, a: int, r: float, s_next: State, done: bool) -> None:
        current = self._q(s, a)
        if done:
            target = r
        else:
            best_next = max(self._q(s_next, a2) for a2 in range(N_ACTIONS))
            target = r + self.gamma * best_next
        self.q[(s, a)] = current + self.alpha * (target - current)

    def decay_epsilon(self) -> None:
        self.epsilon = max(self.epsilon_end, self.epsilon * self.epsilon_decay)

    # ------------------------------------------------------------------
    # Persistence
    # ------------------------------------------------------------------

    def save(self, path: str | Path) -> None:
        p = Path(path)
        p.parent.mkdir(parents=True, exist_ok=True)
        with open(p, "wb") as f:
            pickle.dump(
                {
                    "gamma": self.gamma, "alpha": self.alpha,
                    "epsilon": self.epsilon, "epsilon_end": self.epsilon_end,
                    "epsilon_decay": self.epsilon_decay, "seed": self.seed,
                    "q": self.q,
                },
                f,
            )

    @classmethod
    def load(cls, path: str | Path) -> "QAgent":
        with open(path, "rb") as f:
            data = pickle.load(f)
        agent = cls(
            gamma=data["gamma"], alpha=data["alpha"], epsilon=data["epsilon"],
            epsilon_end=data["epsilon_end"], epsilon_decay=data["epsilon_decay"],
            seed=data["seed"],
        )
        agent.q = data["q"]
        return agent
